### PR TITLE
Some pathing docs

### PR DESCRIPTION
**Pull request by hensldm**
1. Names the two unk values in the `Path` struct
    - I'm not sure on `customValue` but can't really think of anything better. It is used as dayFlags in subs, pathtype for shooting gallery actors, and also seems to get compared to the current waypoint in `EnPr2`.
    - Documents a little for the shooting gallery actors that use `path.customValue`

2. Adds defines for each actor to indicate their NONE_PATH_INDEX (basically -1 to some bit width based on their own params), as well as two defines 1 for the general NONE_PATH_INDEX of just a regular -1, and one for the additional path index from scenes which is `-1` with 8 bits.

### CODE REVIEW

File: src/overlays/actors/ovl_Bg_F40_Block/z_bg_f40_block.c [hunk 1]
```diff
@@ -56,7 +56,7 @@ s32 func_80BC3980(BgF40Block* this, PlayState* play) {
     this->unk_160 = 0;
     this->unk_164 = 0;
 
-    if (BGF40BLOCK_GET_PATH(&this->dyna.actor) != 0x3F) {
+    if (BGF40BLOCK_GET_PATH(&this->dyna.actor) != PATH_INDEX_MAX) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_F40_Block/z_bg_f40_block.c:59]:
```suggestion
    if (BGF40BLOCK_GET_PATH_INDEX(&this->dyna.actor) != PATH_INDEX_MAX) {
```
Very tempted to go through and comment on some consistency things and name some actor unknowns. (I'm seeing `pathId`, `pathnum`, `pathLocation`), would that cleanup be something you're interested in? Can leave it outside the scope of this PR if you'd like and note it down for later

File: src/overlays/actors/ovl_En_Ge2/z_en_ge2.c [hunk 2]
```diff
@@ -228,14 +228,14 @@ void EnGe2_GetNextPath(EnGe2* this, PlayState* play) {
...
-    u8 unk1;
+    u8 nextPathIndex;
 
     this->curPointIndex = 0;
 
     if (GERUDO_PURPLE_GET_PATH(&this->picto.actor) != GERUDO_PURPLE_PATH_NONE) {
         curPath = &play->setupPathList[GERUDO_PURPLE_GET_PATH(&this->picto.actor)];
-        unk1 = curPath->unk1;
-        nextPath = &play->setupPathList[unk1];
+        nextPathIndex = curPath->additionalPathIndex;
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Ge2/z_en_ge2.c:237]: What are your thoughts for `nextPathIndex` vs `additionalPathIndex`? Why go with `additionalPathIndex` over `nextPathIndex` for the main struct?

Comment (hensldm) [reply src/overlays/actors/ovl_En_Ge2/z_en_ge2.c:237]: https://github.com/zeldaret/mm/pull/823#discussion_r890676778
TLDR: Seems similar to the `additionalCsId` so like to mirror it.
(I also got the idea for customValue in that PR as well so thank you :) ).

File: src/overlays/actors/ovl_Bg_F40_Block/z_bg_f40_block.c (hunk 1 above)
Comment (hensldm) [reply src/overlays/actors/ovl_Bg_F40_Block/z_bg_f40_block.c:59]: I don't mind too much doing it. Give me a bit to actually go through and clean them up before you add all the reviews please :).

Comment (hensldm) [reply src/overlays/actors/ovl_Bg_F40_Block/z_bg_f40_block.c]: Ok I think that is it for now.

File: src/overlays/actors/ovl_En_Aob_01/z_en_aob_01.c [hunk 3]
```diff
@@ -110,12 +110,12 @@ void func_809C1158(EnAob01* this, PlayState* play) {
     s32 temp_s0 = ENAOB01_GET_7E00_2(&this->actor);
     s16 i = 0;
 
-    if (temp_s0 != 0x3F) {
+    if (temp_s0 != PATH_INDEX_MAX) {
         do {
-            this->unk_1D8[i] = SubS_GetPathByIndex(play, temp_s0, 0x3F);
-            temp_s0 = this->unk_1D8[i]->unk1;
+            this->unk_1D8[i] = SubS_GetPathByIndex(play, temp_s0, PATH_INDEX_MAX);
+            temp_s0 = this->unk_1D8[i]->additionalPathIndex;
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Aob_01/z_en_aob_01.c:116]:
```suggestion
            pathIndex = this->paths[i]->additionalPathIndex;
```

File: src/overlays/actors/ovl_En_Aob_01/z_en_aob_01.c [hunk 4]
```diff
@@ -110,12 +110,12 @@ void func_809C1158(EnAob01* this, PlayState* play) {
     s32 temp_s0 = ENAOB01_GET_7E00_2(&this->actor);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Aob_01/z_en_aob_01.c:110]:
```suggestion
    s32 pathIndex = ENAOB01_GET_PATH_INDEX(&this->actor);
```

File: src/overlays/actors/ovl_En_Daiku/z_en_daiku.c [hunk 5]
```diff
@@ -134,8 +134,8 @@ void func_8094373C(EnDaiku* this, s32 arg1) {
 }
 
 void func_809437C8(EnDaiku* this) {
-    if ((this->unk_288 != -1) && (this->unk_258 != 0)) {
-        if (!SubS_CopyPointFromPath(this->unk_258, this->unk_25C, &this->unk_26C)) {
+    if ((this->pathIndex != -1) && (this->path != 0)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Daiku/z_en_daiku.c:137]:
```suggestion
    if ((this->pathIndex != -1) && (this->path != NULL)) {
```

File: src/overlays/actors/ovl_En_Daiku/z_en_daiku.c [hunk 6]
```diff
@@ -207,9 +207,9 @@ void func_809438F8(EnDaiku* this, PlayState* play) {
 
         sq = sqrtf(SQ(this->actor.world.pos.x - this->unk_26C.x) + SQ(this->actor.world.pos.z - this->unk_26C.z));
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Daiku/z_en_daiku.c:208]: Not you, but `sq` is incorrect here since the sqrt is taken, meaning `SQ(2.0f)` should just be `4.0f` below

File: src/overlays/actors/ovl_En_Famos/z_en_famos.c [hunk 7]
```diff
@@ -366,7 +366,7 @@ void EnFamos_StillIdle(EnFamos* this, PlayState* play) {
 void EnFamos_SetupPathingIdle(EnFamos* this) {
     if (this->isCalm) {
         this->currentPathNode++;
-        if (this->currentPathNode == this->pathNodeCount) {
+        if (this->currentPathNode == this->pathCount) {
             this->currentPathNode = 0;
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Famos/z_en_famos.c:370]:
```suggestion
        if (this->pathIndex == this->pathCount) {
            this->pathIndex = 0;
```

File: src/overlays/actors/ovl_En_Suttari/z_en_suttari.c [hunk 8]
```diff
@@ -679,7 +679,7 @@ void func_80BABB90(EnSuttari* this, s32 arg1) {
 s32 func_80BABC48(EnSuttari* this, PlayState* play, ScheduleOutput* scheduleOutput) {
     u16 sp26 = SCHEDULE_TIME_NOW;
     u16 phi_a0;
-    u8 sp23 = ENSUTTARI_GET_PATH(&this->actor);
+    u8 sp23 = ENSUTTARI_GET_PATH_INDEX(&this->actor);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Suttari/z_en_suttari.c:682]:
```suggestion
    u8 pathIndex = ENSUTTARI_GET_PATH_INDEX(&this->actor);
```

File: src/overlays/actors/ovl_En_Suttari/z_en_suttari.c [hunk 9]
```diff
@@ -710,7 +710,7 @@ s32 func_80BABC48(EnSuttari* this, PlayState* play, ScheduleOutput* scheduleOutp
 s32 func_80BABDD8(EnSuttari* this, PlayState* play, ScheduleOutput* scheduleOutput) {
     s32 pad;
     EnDoor* sp48;
-    u8 sp47 = ENSUTTARI_GET_PATH(&this->actor);
+    u8 sp47 = ENSUTTARI_GET_PATH_INDEX(&this->actor);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Suttari/z_en_suttari.c:713]:
```suggestion
    u8 pathIndex = ENSUTTARI_GET_PATH_INDEX(&this->actor);
```

File: src/overlays/actors/ovl_En_Aob_01/z_en_aob_01.c (hunk 3 above)
Comment (hensldm) [reply src/overlays/actors/ovl_En_Aob_01/z_en_aob_01.c:116]: Aob01 is weird. It isn't her path, but related to the dogs. I'd rather wait for more docs around the whole thing which darkeye has been working on (https://discord.com/channels/688807550715560050/873211000122384415/1080652253498716190)

File: src/overlays/actors/ovl_En_Aob_01/z_en_aob_01.c (hunk 4 above)
Comment (hensldm) [reply src/overlays/actors/ovl_En_Aob_01/z_en_aob_01.c:110]: Same as above

File: src/overlays/actors/ovl_En_Daiku/z_en_daiku.c (hunk 6 above)
Comment (hensldm) [reply src/overlays/actors/ovl_En_Daiku/z_en_daiku.c:208]: Renamed the variables as well.

File: src/overlays/actors/ovl_En_Famos/z_en_famos.c (hunk 7 above)
Comment (hensldm) [reply src/overlays/actors/ovl_En_Famos/z_en_famos.c:370]: Not really the `pathIndex`, it is the `waypointIndex` (current point on the path)

File: src/overlays/actors/ovl_En_Bom_Bowl_Man/z_en_bom_bowl_man.c [hunk 10]
```diff
@@ -156,7 +156,7 @@ void func_809C4B50(EnBomBowlMan* this) {
 }
 
 void func_809C4B6C(EnBomBowlMan* this) {
-    if ((this->unk_29A != ENBOMBOWLMAN_FF00_MINUS1) && (this->path != NULL)) {
+    if ((this->pathIndex != -1) && (this->path != NULL)) {
```
[the remark this answers is not in the document]
Comment (isghj5) [reply src/overlays/actors/ovl_En_Bom_Bowl_Man/z_en_bom_bowl_man.c:159]: Isn't FF already defined?

[the remark this answers is not in the document]
Comment (engineer124) [reply src/overlays/actors/ovl_En_Bom_Bowl_Man/z_en_bom_bowl_man.c:159]: Unfortunately, `pathIndex` is most commonly `s16` so it doesn't overlap. The occasional `s32` too it seems.

[the remark this answers is not in the document]
Comment (isghj5) [reply src/overlays/actors/ovl_En_Bom_Bowl_Man/z_en_bom_bowl_man.c:159]: I'm suspicious that so many paths can exist at once

File: src/overlays/actors/ovl_En_Ge2/z_en_ge2.c (hunk 2 above)
Comment (mzxrules) [reply src/overlays/actors/ovl_En_Ge2/z_en_ge2.c:237]: I haven't looked at the z_eventmgr in detail to know why it's called additionalCs there, but here `additionalPathIndex` references the next element in a linked list of `Path`s, so I don't understand the argument for not using `next` here.

File: src/overlays/actors/ovl_Obj_Toge/z_obj_toge.c [hunk 11]
```diff
@@ -131,12 +131,12 @@ void ObjToge_Init(Actor* thisx, PlayState* play) {
 
     Collider_InitCylinder(play, &this->collider);
 
-    if (OBJTOGE_GET_PATH(thisx) == 0xFF) {
+    if (OBJTOGE_GET_PATH_INDEX(thisx) == PATH_INDEX_END) {
```
[the remark this answers is not in the document]
Comment (hensldm) [reply src/overlays/actors/ovl_Obj_Toge/z_obj_toge.c]: This should be addressed with the new implementation.

File: include/z64scene.h [hunk 12]
```diff
@@ -461,10 +461,15 @@ typedef struct {
     /* 0x8 */ u32 dListEnd;
 } MeshHeader2; // size = 0xC 
 
+#define PATH_INDEX_END 0xFF
+#define PATH_INDEX_MAX 0x3F
+#define PATH_INDEX_MAX_ALT 0x1F
+#define PATH_INDEX_NONE -1
```
[the remark this answers is not in the document]
Comment (hensldm) [reply include/z64scene.h]: This should be addressed with the new implementation.

File: src/overlays/actors/ovl_En_Ge2/z_en_ge2.c (hunk 2 above)
Comment (hensldm) [reply src/overlays/actors/ovl_En_Ge2/z_en_ge2.c:237]: Tbh I don't really have a argument against `next`, just that it looks to be the same kind of implementation (pretty sure it is linked lists) so would like to mirror the naming to reflect this.

Comment (mzxrules) [reply src/overlays/actors/ovl_En_Ge2/z_en_ge2.c:237]: Well, considering that three different people have suggested calling it nextPathIndex and you say that `additionalCsId` is implemented similarly, I have to ask why `additionalCsId` isn't `nextCsId`?

Comment (engineer124) [reply src/overlays/actors/ovl_En_Ge2/z_en_ge2.c:237]: `nextCsId` makes it sound like it's consecutive cutscenes played back-to-back which isn't always the case. It's a linked list, but `additionalCsId` makes it sound more optional which feels better to describe how the linked list in cutscenes are used

File: src/overlays/actors/ovl_En_Suttari/z_en_suttari.c [hunk 13]
```diff
@@ -516,8 +516,8 @@ void func_80BAB434(EnSuttari* this) {
 }
 
 void EnSuttari_GetPaths(EnSuttari* this, PlayState* play) {
-    this->paths[0] = SubS_GetPathByIndex(play, ENSUTTARI_GET_PATH_INDEX(&this->actor), PATH_INDEX_MAX);
-    this->paths[1] = SubS_GetPathByIndex(play, this->paths[0]->additionalPathIndex, PATH_INDEX_MAX);
+    this->paths[0] = SubS_GetPathByIndex(play, ENSUTTARI_GET_PATH_INDEX(&this->actor), ENSUTTARI_PATH_INDEX_NONE);
+    this->paths[1] = SubS_GetPathByIndex(play, this->paths[0]->additionalPathIndex, ENSUTTARI_PATH_INDEX_NONE);
```
Comment (mzxrules) [inline src/overlays/actors/ovl_En_Suttari/z_en_suttari.c:520]:
```suggestion
    this->paths[0] = SubS_GetPathByIndex(play, ENSUTTARI_GET_PATH_INDEX(&this->actor), ENSUTTARI_PATH_INDEX_NONE);
    //! @bug PATH_INDEX_NONE should be (u8)PATH_INDEX_NONE
    this->paths[1] = SubS_GetPathByIndex(play, this->paths[0]->additionalPathIndex, ENSUTTARI_PATH_INDEX_NONE);
```

However, it'd be cleaner to keep a `PATH_INDEX_END` defined as 0xFF to avoid, I meant that you needed to keep the end of the list of `Path`s defined as 0xFF

File: src/overlays/actors/ovl_En_Ge2/z_en_ge2.c (hunk 2 above)
Comment (hensldm) [reply src/overlays/actors/ovl_En_Ge2/z_en_ge2.c:237]: For what it is worth. This seems to be similar to `paths` sometimes the `next` path isn't the actual next path to be used. It just happens to be the next `path` in the linked list.

File: src/overlays/actors/ovl_En_Bom_Bowl_Man/z_en_bom_bowl_man.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (hensldm) [reply src/overlays/actors/ovl_En_Bom_Bowl_Man/z_en_bom_bowl_man.c]: From what I can tell, it isn't really expected those are the number of paths. It is more that the `NONE_INDEX` is some version of  `-1` i.e all bits are set. `0xFF` is just `-1` for 8 bits.

File: include/z64scene.h [hunk 14]
```diff
@@ -462,6 +462,7 @@ typedef struct {
 } MeshHeader2; // size = 0xC 
 
 #define PATH_INDEX_NONE -1
+#define ADDITIONAL_PATH_INDEX_NONE (u8)PATH_INDEX_NONE
```
Comment (mzxrules) [inline include/z64scene.h:465]: are we really tying every single "_PATH_INDEX_NONE" to a reference to `PATH_INDEX_NONE`

Comment (hensldm) [reply include/z64scene.h:465]: I'll admit maybe I have been a bit carried away with correlating them all with `PATH_INDEX_NONE`. However I do think it is still important/a good idea to show these are all just different `-1`s expressed in different widths. With that in mind, I suggest change them all to be literals, but instead of expressing them as hex literals, `0x1F`, use `-1` with ands/casts, `((-1) & 0x1F)`. More or less this would just replace the `PATH_INDEX_NONE`s with the actual literal `-1`.

Comment (hensldm) [reply include/z64scene.h]: I did eventually just decouple everything and just used the literal hex value i.e (0x3F).

**Comment by hensldm** [comment]
Small note but I had missed some actor clean up regarding path indices, so added them in the latest commit.

**Review by AngheloAlf — APPROVED** [review]
LGTM

**Review by engineer124 — COMMENTED** [review]
Final cleanup comments

**Review by engineer124 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Hidden_Nuts/z_en_hidden_nuts.h [hunk 15]
```diff
@@ -8,7 +8,10 @@ struct EnHiddenNuts;
 typedef void (*EnHiddenNutsActionFunc)(struct EnHiddenNuts*, PlayState*);
 
 #define ENHIDDENNUTS_GET_SWITCHFLAG(thisx) ((thisx)->params & 0x7F)
-#define ENHIDDENNUTS_GET_F80(thisx) (((thisx)->params >> 7) & 0x1F)
+#define ENHIDDENNUTS_GET_PATH_INDEX(thisx) (((thisx)->params >> 7) & 0x1F)
+
+#define ENHIDDENNUTS_PATH_INDEX_NONE 0x1F
+#define ENHIDDENNUTS_PATH_INDEX_NONE_ALT 0x3F
```
Comment (isghj5) [inline src/overlays/actors/ovl_En_Hidden_Nuts/z_en_hidden_nuts.h:14]: man I knew it was different for different actors... but even in the same actor?

File: src/overlays/actors/ovl_En_Rg/z_en_rg.c [hunk 16]
```diff
@@ -741,7 +741,8 @@ void EnRg_Init(Actor* thisx, PlayState* play) {
 
         Effect_Add(play, &this->unk_340, EFFECT_TIRE_MARK, 0, 0, &D_80BF59F0);
 
-        this->path = SubS_GetDayDependentPath(play, ENRG_GET_7F80(&this->actor), 255, &this->unk_33C);
+        this->path =
+            SubS_GetDayDependentPath(play, ENRG_GET_PATH_INDEX(&this->actor), ENRG_PATH_INDEX_NONE, &this->unk_33C);
```
Comment (isghj5) [inline src/overlays/actors/ovl_En_Rg/z_en_rg.c:745]: Formater... why

**Review by isghj5 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Hidden_Nuts/z_en_hidden_nuts.h (hunk 15 above)
Comment (hensldm) [reply src/overlays/actors/ovl_En_Hidden_Nuts/z_en_hidden_nuts.h:14]: Yeah, looks like a copy paste error. `ENHIDDENNUTS_PATH_INDEX_NONE` should be the real one, but use the `ALT` version in the SubS function. However it is harmless since they have the check just before, but probably should have used `0x1F` again.

**Review by tom-overton — APPROVED** [review]
